#cd /s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA
#! which R  source .bashrc   /s/f/shiyi/miniconda3/envs/SCST_2/lib/R/bin/R

```{r setup}
library(spacexr)
library(Matrix)
library(Seurat)
library(dplyr)
```

### Single-Cell Reference

In order to run RCTD, the first step is to process the single cell reference. The reference is created using the RCTD `Reference` constructor function, which requires 3 parameters: <br/>
1. counts: A matrix (or dgCmatrix) representing Digital Gene Expression (DGE). Rownames should be genes and colnames represent barcodes/cell names. Counts should be untransformed count-level data. <br/>
2. cell_types: 	
A named (by cell barcode) factor of cell type for each cell. The 'levels' of the factor would be the possible cell type identities. <br/>
3. nUMI: 	
Optional, a named (by cell barcode) list of total counts or UMI's appearing at each pixel. If not provided, nUMI will be assumed to be the total counts appearing on each pixel. <br/>

The reference may come from various data types, but it needs to be loaded into the R envirnoment. Here, we emphasize that our protocol for loading in the counts, cell_types, and nUMI obejects does not need to be the way that you load your objects into R. In this case our reference is stored in the 'Reference/Vignette' folder as two csv files:<br/>
1. meta_data.csv: a CSV file (with 3 columns, with headers "barcode", "cluster", and "nUMI") containing the nUMI and cell type assignment for each cell.<br/>
 <br/>
2. dge.csv: a Digital Gene Expression (DGE) (barcodes by gene counts) CSV file in the standard 10x format. <br/>

```{r scRNA: use desc celltype desc_0.2}
### Load in/preprocess your data, this might vary based on your file type
#counts=genes*barcodes/cell
#meta_data=headers "barcode", "cluster", and "nUMI"
counts <-Matrix::readMM("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/1_mydesc/sc_combine_counts_sparse.mtx")
dim(counts) #3348 40532 #12846 40532
var = read.csv("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/1_mydesc/dge/var.csv")
head(var)
rownames(counts)=make.unique(var$X, sep = "_")
meta_data = read.csv("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/1_mydesc/dge/obs.csv") # load in meta_data (barcodes, clusters, and nUMI)
head(meta_data)
colnames(counts)<- meta_data$X
head(colnames(counts))
head(rownames(counts))
cell_types <- meta_data$desc_0.2; names(cell_types) <- meta_data$X # create cell_types named list
cell_types <- as.factor(cell_types) # convert to factor data type
reference <- Reference(counts, cell_types)  #, nUMI),n_max_cells = 20000

## Examine reference object (optional)
print(dim(reference@counts)) #observe Digital Gene Expression matrix
table(reference@cell_types) #number of occurences for each cell type
"""0     1     2     3     4     5     6     7     8     9    10    11    12 
10000  3679  2770  3017  1892  1801  1546  1849  1453  1278  1377  1222   892 
   13    14    15    16    17    18    19    20    21    22    23    24    25 
  595   835   702   597   428   383   469   439   287   279   264   204   210 
   26    27    28    29 
  395   100    79    30 
 """
## Save RDS object (optional)
saveRDS(reference, '/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/mySCRef.rds') #annotation_2减少celltype个数
```



### Spatial Transcriptomics data
Next, we will load the Spatial Transcriptomics data into a `SpatialRNA` object. Similarly to the reference, we will first load the data into R, and then pass the data into the RCTD `SpatialRNA` constructor function, which requires 3 parameters: <br/>
1. coords: A numeric data.frame (or matrix) representing the spatial pixel locations. rownames are barcodes/pixel names, and there should be two columns for 'x' and for 'y'.<br/>
2. counts: A matrix (or dgCmatrix) representing Digital Gene Expression (DGE). Rownames should be genes and colnames represent barcodes/pixel names. Counts should be untransformed count-level data. <br/>
3. nUMI: 	
Optional, a named (by pixel barcode) list of total counts or UMI's appearing at each pixel. If not provided, nUMI will be assumed to be the total counts appearing on each pixel. <br/>

These elements can be constructed differently, depending on how you store your spatial transcriptomic data. In our case (one of many ways to do it), our (Slide-seq) data is stored in the 'SpatialRNA/Vignette' file as two csv files: <br/>
1. BeadLocationsForR.csv: a CSV file (with 3 columns, with headers "barcodes", "xcoord", and "ycoord") containing the spatial locations of the pixels. <br/>
2. MappedDGEForR.csv: a DGE (gene counts by barcodes) CSV file. Represents raw counts at each pixel. <br/>

#stdata:/s/f/shiyi/ST/Data/test/CELL2LOCATION.R
#ST8059048 = sc.read_visium(path='/s/f/shiyi/ST/Data/cell2location/mouse_brain_visium_wo_cloupe_data/mouse_brain_visium_wo_cloupe_data/rawdata/ST8059048',count_file='filtered_feature_bc_matrix.h5', load_images=True)

```{r SpatialRNA, results = 'hide', fig.height = 6, fig.width = 6}
st_combine=readRDS("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/under_sample/st_subsetRCTD/st_combine.rds")#3348 features across 2987 samples
counts = st_combine@assays$Spatial@counts   # 3348 2987
coords= st_combine@images$slice1@coordinates[4:5]    #2987    2
head(rownames(counts))
head(rownames(coords))
nUMI <- colSums(counts) # In this case, total counts per pixel is nUMI

### Create SpatialRNA object
puck <- SpatialRNA(coords, counts, nUMI)

## Examine SpatialRNA object (optional)
print(dim(puck@counts)) # observe Digital Gene Expression matrix  # 2000 2987
hist(log(puck@nUMI,2)) # histogram of log_2 nUMI
print(head(puck@coords)) # start of coordinate data.frame
barcodes <- colnames(puck@counts) # pixels to be used (a list of barcode names). 

saveRDS(puck,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/myRCTDspatial.rds")

```


### Creating RCTD Object
### Running RCTD
Now, we are ready to run RCTD, using the `run.RCTD` function. This function is equivalent to sequentially running the functions `fitBulk`, `choose_sigma_c`, and `fitPixels`. The `doublet_mode` argument sets whether RCTD will be run in 'doublet mode' (at most 1-2 cell types per pixel), 'full mode' (no restrictions on number of cell types), or 'multi mode' (finitely many cell types per pixel, e.g. 3 or 4).

```{r DEgenes}
#spatial: sc: merge common genes + desc
myRCTD <- create.RCTD(puck, reference, max_cores = 1,CELL_MIN_INSTANCE = 10)
myRCTD_1 <- run.RCTD(myRCTD, doublet_mode = 'doublet')
saveRDS(myRCTD_1,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/myRCTD_doublet.rds") 


```


```{r }
myRCTD_1=readRDS("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/myRCTD_doublet.rds")  #doublet,hvg_st=2000
results <- myRCTD_1@results
norm_weights = normalize_weights(results$weights) # 2982   30
#aa=rowSums(norm_weights)
mymax=apply(norm_weights, 1,max)
norm_weights_1=cbind(norm_weights,mymax)  #加入max列，筛选max>n的行
dim(norm_weights_1)
newdata=norm_weights_1[norm_weights_1[,'mymax']>=mean(mymax),]  
dim(newdata)  #1514   31
max_1=apply(newdata, 1, function(t) colnames(newdata)[which.max(t)])  #输出最大概率对应的列名-celltype
table(max_1)    
length(max_1) # 1514
write.csv(max_1, file = "/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/spots_celltype_max.csv")
"""0   1  13  14  15  18  19   2  21  22  24  27  29   4   6   7 
180  38 906   7  23   2  13  11  19   1  18   1   3 259  30   3 
"""
```

```{r:subset rctd}
library(SeuratDisk)
library(SeuratData)

#st_combine=readRDS("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/under_sample/st_subsetRCTD/st_combine.rds")
max_1 = read.csv("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/spots_celltype_max.csv") #挑选出单一细胞类型的spots
table(max_1) 
length(max_1) #1514
head(max_1$barcode)
st_combine_subset=subset(st_combine,cells=max_1$barcode,slot="counts" ) 
st_combine_subset@meta.data$desc_0.2=max_1$x#3348 features across 1514 samples
saveRDS(st_combine_subset,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/st_combine_subset.rds")
SaveH5Seurat(st_combine_subset, filename = "/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/st_combine_subset.h5Seurat")
Convert("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/2_RCTD/st_combine_subset.h5Seurat", dest = "h5ad")
```



###NO#######################################################
```{r spots}
#newdata<-subset(norm_weights_1, norm_weights_1[,'mymax'] >=0.8,select=c(colnames(norm_weights)))  #"numeric"
#re1 = norm_weights_1 %>% filter(norm_weights_1[,'mymax']>0.8)  #Error in UseMethod("filter") : no applicable method for 'filter' applied to an object of class "c('dgeMatrix',
#max=max(norm_weights)
#max=apply(norm_weights, 1,function(t) which(which.max(t)>0.8,arr.ind=TRUE))
#max=apply(norm_weights, 1,function(t) filter((max(t)>0.8)==FALSE))

```

```{r preprocess_scRNA}
'''
#/s/f/shiyi/ST/Data/cell2location/sc.h5ad
#/s/f/shiyi/ST/Data/cell2location/scdata
dat=readRDS("/s/f/shiyi/ST/Data/cell2location/data/filter_raw.rds")
dat
head(colnames(dat))
tail(colnames(dat))
counts=dat$RNA@counts
write.csv(counts, file = "dge.csv", append = FALSE)
meta_data=dat$meta.data
'''
```
```{r scRNA_anno2:不用此数据}
### Load in/preprocess your data, this might vary based on your file type
counts <-Matrix::readMM("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/1_mydesc/counts_sparse.mtx")
dim(counts) #12846 40532
var = read.csv("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/1_mydesc/dge_anno2/var.csv")
rownames(counts)=make.unique(var$SYMBOL, sep = "_")
meta_data = read.csv("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/1_mydesc/dge_anno2/obs.csv") # load in meta_data (barcodes, clusters, and nUMI)
#meta_data = read.csv("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/0727/1_mydesc/meta_data_anno2.csv") # load in meta_data (barcodes, clusters, and nUMI)
colnames(counts)<- meta_data$barcode
head(colnames(counts))
head(rownames(counts))
cell_types <- meta_data$annotation_2; names(cell_types) <- meta_data$barcode # create cell_types named list
cell_types <- as.factor(cell_types) # convert to factor data type
table(cell_types)
reference <- Reference(counts, cell_types,n_max_cells = 20000)  
## Examine reference object (optional)
print(dim(reference@counts)) #observe Digital Gene Expression matrix
table(reference@cell_types) #number of occurences for each cell type
"""Astro  Endo   Ext   Inh  LowQ Micro    Nb Oligo   OPC   Unk 
 3013    22 13309  8471   375  1218   428 11432  1315   949 """
## Save RDS object (optional)
saveRDS(reference, '/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/mySCRef_anno2.rds') #annotation_2减少celltype个数
#saveRDS(reference, '/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/mySCRef.rds')
#saveRDS(reference, '/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/mySCRef_1.rds') #
#saveRDS(reference, '/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/mySCRef_2.rds') #annotation_2减少celltype个数

```

```{r DEgenes}
#spatial:hvg2000; sc:不处理
puck=readRDS("/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/myRCTDspatial_hvg.rds")
reference=readRDS('/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/mySCRef_anno2.rds')
myRCTD <- create.RCTD(puck, reference, max_cores = 1,CELL_MIN_INSTANCE = 10)
myRCTD_1 <- run.RCTD(myRCTD, doublet_mode = 'doublet')
saveRDS(myRCTD_1,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/myRCTD_doublet_0419.rds")  #2_1:reference=20000; 

#saveRDS(myRCTD_1,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/myRCTD_doublet_anno2.rds")
#myRCTD_2 <- run.RCTD(myRCTD, doublet_mode = 'full')
#saveRDS(myRCTD_1,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/myRCTD_doublet_1.rds")  #doublet,hvg_st=2000
#saveRDS(myRCTD_1,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/myRCTD_doublet_2.rds")  #2:dicrease sc_celltype+doublet,hvg_st=2000
#saveRDS(myRCTD_1,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/myRCTD_doublet_2_1.rds")  #2_1:reference=20000; 
#saveRDS(myRCTD_2,"/s/f/shiyi/ST/hierarchy_dense_net/test1029/mynetwork/mynetwork_scstmerge_deconvolution/MY_RCTD/MYDATA/myRCTD_full_2.rds")  #full,dicrease sc_celltype, hvg_st=2000,
```

#doublet:2_1:reference=20000; 2:dicrease sc_celltype,doublet,hvg_st=2000
#full:2
```{r weights of each spot}
results <- myRCTD_1@results
norm_weights = normalize_weights(results$weights) #  2978   10
aa=rowSums(norm_weights)  #=1
#write.csv(norm_weights, file = "spots_celltype.csv", append = FALSE)
#write.csv(norm_weights, file = "spots_celltype_2.csv", append = FALSE)
write.csv(norm_weights, file = "spots_celltype_2_1.csv", append = FALSE)

#过滤大于0.8的概率，输出列名
mymax=apply(norm_weights, 1,max)
norm_weights_1=cbind(norm_weights,mymax)  #加入max列，筛选max>n的行
dim(norm_weights_1)
newdata=norm_weights_1[norm_weights_1[,'mymax']>=0.6,]  #anno2: 908  11#0.6：316  60 ；0.8：14 60
dim(newdata)  
max_1=apply(newdata, 1, function(t) colnames(newdata)[which.max(t)])  #输出最大概率对应的列名-celltype
table(max_1)    
"""> table(max_1)   
max_1
  Ext   Inh    Nb Oligo   Unk 
    1     7     1   296   603 
有问题！！
2_1
max_1
  Ext   Inh    Nb Oligo   Unk 
    1     6     1   296   603 

"""
#write.csv(max_1, file = "spots_celltype_max_2.csv") #挑选出单一细胞类型的spots
write.csv(max_1, file = "spots_celltype_max_2_1.csv") #挑选出单一细胞类型的spots

max_1 = read.csv("spots_celltype_max.csv") #挑选出单一细胞类型的spots
table(max_1$x)  


spatialRNA <- myRCTD_1@spatialRNA
weights_doublet=results$weights_doublet #2978    2
head(weights_doublet)
results_df=results$results_df
dim(results_df) #2978    9
dim(results_df[results_df$spot_class=='singlet',])  #592   9
```


```{r weights of each spot}
results <- myRCTD_2@results
norm_weights = normalize_weights(results$weights) # 2978   59
aa=rowSums(norm_weights)  #=1
#write.csv(norm_weights, file = "spots_celltype.csv", append = FALSE)
write.csv(norm_weights, file = "spots_celltype_2_full.csv", append = FALSE)
#过滤大于0.8的概率，输出列名
mymax=apply(norm_weights, 1,max)
norm_weights_1=cbind(norm_weights,mymax)  #加入max列，筛选max>n的行
dim(norm_weights_1)
newdata=norm_weights_1[norm_weights_1[,'mymax']>=0.6,]  #anno2: 908  11#0.6：316  60 ；0.8：14 60
dim(newdata)  #909  11
max_1=apply(newdata, 1, function(t) colnames(newdata)[which.max(t)])  #输出最大概率对应的列名-celltype
table(max_1)    
"""> table(max_1)   
max_1
  Ext   Inh    Nb Oligo   Unk 
    1     8     1   296   603 
有问题！！
"""
write.csv(max_1, file = "spots_celltype_max_2_full.csv") #挑选出单一细胞类型的spots



```

